system/security/shared/caching.cpp

```diff
@@ -585,6 +562,46 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
         return true;
     }
 
+    // If the cache is not ready, then an initial fill is required. Acquire the cache RW lock and make sure no other thread has
+    // tried to fill the cache. Once filled, start the fill thread and wait for it to be ready.
+    if (!m_fileScopeCacheReady)
+    {
+        WriteLockBlock writeLock(m_scopesRWLock);
+
+        if (!m_fileScopeCacheReady)
+        {
+            IArrayOf<ISecResource> scopes;
+            aindex_t count = m_secMgr->getManagedScopeTree(RT_FILE_SCOPE, nullptr, scopes);
+            removeAllManagedFileScopes();
```

Comment:
This should be commoned up with the code in the main thread.  It likely needs a flag to indicate the lock is already obtained. Extending the WriteLockBlock class would make this easier.